### PR TITLE
Add topographic unit as new ELM subgrid level

### PR DESCRIPTION
**Pull request by thorntonpe**
Introduces a new level, the topographic unit, in the ELM subgrid hierarchy. Topographic units are introduced belwo the gridcell level, and above the landunit level. The functionality is tested to produce BFB results when the max_topounits variable is set to 1. This inserts a single topographic unit on every gridcell, with a weight of 1.0, and exercises all model operations through the new level. The functionality is also tested to produce roundoff-level differences for max_topounits > 1.  In this case, multiple topographic units are created for each gridcell, with equal weights on the gridcell, and each having an exact copy of the original sub-grid structure implemented on each of the gridcell's unique topounits.  This functionality is required for the subsequent introduction of variable numbers of topographic units per gridcell, informed by new surface datasets. 
A known issue with this new capability is that it produces larger-than-roundoff differences for cases with the combined conditions: max_topounits>1, using dynamic PFTs through the land use dataset drivers, and dynamic rooting depth turned on. Since other issues have been identified with use of dynamic rooting depth, and since, as a result, that capability is currently turned off by default in our test suites, this is not seen as a critical issue for the topographic unit capability. Further investigation of the dynamic rooting algorithm interaction with subgrid scheme is warranted.
A second known issue is that introduction of this capability probably breaks the functionality of the dynamic land ice component (glc_mec). That component is not currently supported by E3SM testing, and a separate development effort is underway to revive that model component. A path toward solution seems clear, and in consultation with the dynamic glacier developers it was decided that it is acceptable to temporarily break that capability to move the topounits development process forward.
This new topographic unit capability and the associated changes in data structures are one component of the ongoing land model data architecture v1/v2 development effort.

[BFB] with max_topounits=1

### CODE REVIEW

File: components/clm/src/main/topounit_varcon.F90 [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  save
+  private
+  
+  !------------------------------------------------------------------
+  ! Initialize topounit type constants
+  !------------------------------------------------------------------
+  
+  integer, parameter, public :: max_topounits  = 1 ! maximum number of topounits per gridcell
+
+end module topounit_varcon
```
[the remark this answers is not in the document]
Comment (thorntonpe) [reply components/clm/src/main/topounit_varcon.F90]: Fixed - these were left-over CR-LF characters from a round of editing on my windows platform.

File: components/clm/src/main/subgridWeightsMod.F90 [hunk 2]
```diff
@@ -482,16 +497,16 @@ real(r8) :: weight  ! function result
   end function get_landunit_weight
 
   !-----------------------------------------------------------------------
-  subroutine set_landunit_weight(g, ltype, weight)
+  subroutine set_landunit_weight(t, ltype, weight)
```
[the remark this answers is not in the document]
Comment (thorntonpe) [reply components/clm/src/main/subgridWeightsMod.F90]: As for the "get"  version of this comment: I agree in general, and I would like to come back to make this change later. The reason is that the function is used in a couple other places in the code, for dynamic crops and dynamic land ice, which are known to conflict with this PR. Making the change in those places would also be confusing. As a temporary compromise I improved the commenting within this function to make it clear that the weights are relative to the topounit.

File: components/clm/src/main/lnd2atmMod.F90 [hunk 3]
```diff
@@ -101,6 +119,40 @@ subroutine lnd2atm_minimal(bounds, &
          lnd2atm_vars%eflx_lwrad_out_grc      (bounds%begg:bounds%endg), &
          p2c_scale_type='unity', c2l_scale_type= 'urbanf', l2g_scale_type='unity')
 
+! #define SUBGRID_DEBUG
```
[the remark this answers is not in the document]
Comment (thorntonpe) [reply components/clm/src/main/lnd2atmMod.F90]: Debugging code removed in new commit.

File: components/clm/src/main/initSubgridMod.F90 [hunk 4]
```diff
@@ -120,27 +122,65 @@ subroutine clm_ptrs_compdown(bounds)
...
           call endrun(decomp_index=l, clmlevel=namel, msg=errMsg(__FILE__, __LINE__))
        end if
 
        if (grc_pp%landunit_indices(ltype, curg) == ispval) then
           grc_pp%landunit_indices(ltype, curg) = l
        else
-          write(iulog,*) 'clm_ptrs_compdown ERROR: This landunit type has already been set for this gridcell'
-          write(iulog,*) 'l, ltype, curg = ', l, ltype, curg
-          call endrun(decomp_index=l, clmlevel=namel, msg=errMsg(__FILE__, __LINE__))
+          ! write(iulog,*) 'clm_ptrs_compdown ERROR: This landunit type has already been set for this gridcell'
```
[the remark this answers is not in the document]
Comment (thorntonpe) [reply components/clm/src/main/initSubgridMod.F90]: If block added in new commit.

File: components/clm/src/main/subgridWeightsMod.F90 [hunk 5]
```diff
@@ -453,17 +468,17 @@ logical function is_active_p(p)
   end function is_active_p
 
   !-----------------------------------------------------------------------
-  function get_landunit_weight(g, ltype) result(weight)
+  function get_landunit_weight(t, ltype) result(weight)
```
[the remark this answers is not in the document]
Comment (bishtgautam) [reply components/clm/src/main/subgridWeightsMod.F90]: Ok, sounds good.

File: components/clm/src/main/subgridWeightsMod.F90 (hunk 2 above)
[the remark this answers is not in the document]
Comment (bishtgautam) [reply components/clm/src/main/subgridWeightsMod.F90]: Ok, sounds good.

**Comment by bishtgautam** [comment]
@thorntonpe, I rebased my DLU branches off your branch and there was a single trivial conflict. Since my DLU PRs won't be merged to master till v1.1 is created, @jqyin should feel to merge this PR at the next available slot after @tktesfa2014 completes the review.

**Comment by thorntonpe** [comment]
@bishtgautam , thanks for the review and comments, and for the test rebase with DLU branches.

**Comment by tktesfa2014** [comment]
@thorntonpe I have reviewed the PR. Looks good. Thanks!

**Comment by jqyin** [comment]
Re-merge to fix SMS.f09_g16_a.IGCLM45_MLI test

**Comment by jqyin** [comment]
Merged #2357

**Pull request merged**